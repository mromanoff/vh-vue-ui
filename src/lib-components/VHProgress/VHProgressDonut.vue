<template>
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 170 170"
    class="Progress"
    :class="computedSize"
  >
    <g transform="rotate(90 85 85)">
      <circle
        class="Progress-track"
        stroke-width="22"
        fill="none"
        cx="85"
        cy="85"
        :r="radius"
      />
    </g>
    <g transform="rotate(90 85 85)">
      <circle
        class="Progress-bar"
        :class="computedColor"
        stroke-width="22"
        :stroke-dasharray="strokeDasharray"
        stroke-linecap="butt"
        fill="none"
        cx="85"
        cy="85"
        :r="radius"
      />
    </g>
    <text
      class="Progress-label"
      x="86"
      y="106"
      text-anchor="middle"
    >
      {{ label }}
    </text>
  </svg>
</template>

<script>
/*
Example:
Radius = 85
Percentage to fill = 25
π = 3.14159
Circumference = 2 * π * Radius = 534
Stroke length = (Circumference / 100) * Percentage to fill
*/
export default {
  name: 'VhProgressDonut',
  props: {
    /**
     *  One of these colors: orange, green, blue
     */
    color: {
      type: String,
      default: 'green',
      validator: value => {
        return value.match(/(green|orange|blue)/);
      },
    },
    /**
     * One of these sizes: small, medium, large
     */
    size: {
      type: String,
      default: 'medium',
      validator: value => {
        return value.match(/(small|medium|large)/);
      },
    },
    /**
     * value number from 0 to 100
     */
    value: {
      type: [Number, String],
      default: 0,
    },
    /**
     * label: 4.5 etc.
     */
    label: {
      type: String,
      default: '',
    },
  },

  data() {
    return {
      radius: 70,
      Pi: 3.14159,
    };
  },
  computed: {
    computedSize() {
      return {
        [`Progress--${this.size}`]: this.size,
      };
    },

    computedColor() {
      return {
        [`Progress-bar--${this.color}`]: this.color,
      };
    },

    circumference() {
      return 2 * this.Pi * this.radius;
    },

    strokeLength() {
      return (this.circumference / 100) * this.normalizedValue;
    },

    strokeDasharray() {
      return `${this.strokeLength}, ${this.circumference}`;
    },

    computedWidth() {
      return `${this.normalizedValue}%`;
    },

    normalizedValue() {
      if (this.value < 0) {
        return 0;
      }
      if (this.value > 100) {
        return 100;
      }
      return parseFloat(this.value);
    },
  },
};
</script>

<style scoped>
.Progress {
  display: inline-block;
  font-weight: var(--font-weight--normal);
}

.Progress-track {
  stroke: var(--color--grey200);
}

.Progress-bar {
  height: 100%;
  width: 0;
}

.Progress-label {
  fill: var(--text-color);
  font-size: 60px;
}

/* donut sizes */
.Progress--small {
  height: 36px;
  width: 36px;
}

.Progress--medium {
  height: 46px;
  width: 46px;
}

.Progress--large {
  height: 64px;
  width: 64px;
}

/* bar colors */
.Progress-bar--green {
  stroke: var(--color--green700);
}

.Progress-bar--orange {
  stroke: var(--color--orange500);
}

.Progress-bar--blue {
  stroke: var(--color--blue500);
}
</style>
