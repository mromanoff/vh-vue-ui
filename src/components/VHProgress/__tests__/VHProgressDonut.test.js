import { shallowMount } from '@vue/test-utils';
import VHProgressDonut from '../VHProgressDonut.vue';

describe('VHProgressDonut.vue', () => {
  let wrapper;

  it('renders default VHProgressDonut', () => {
    wrapper = shallowMount(VHProgressDonut);
    expect(wrapper.props().color).toBe('green');
    expect(wrapper.props().size).toBe('medium');
    expect(wrapper.props().value).toBe(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders modified VHProgressDonut blue large', () => {
    wrapper = shallowMount(VHProgressDonut, {
      propsData: {
        color: 'blue',
        size: 'large',
      },
    });
    expect(wrapper.props().color).toBe('blue');
    expect(wrapper.props().size).toBe('large');
    expect(wrapper.props().value).toBe(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders modified VHProgressDonut green small', () => {
    wrapper = shallowMount(VHProgressDonut, {
      propsData: {
        color: 'green',
        size: 'small',
        value: 10,
      },
    });
    expect(wrapper.props().color).toBe('green');
    expect(wrapper.props().size).toBe('small');
    expect(wrapper.props().value).toBe(10);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders modified VHProgressDonut blue small', () => {
    wrapper = shallowMount(VHProgressDonut, {
      propsData: {
        color: 'blue',
        size: 'medium',
        value: 1,
      },
    });
    expect(wrapper.props().color).toBe('blue');
    expect(wrapper.props().size).toBe('medium');
    expect(wrapper.props().value).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });
});
