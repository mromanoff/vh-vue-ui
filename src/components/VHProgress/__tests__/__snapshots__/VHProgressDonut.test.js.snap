// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VHProgressDonut.vue renders default VHProgressDonut 1`] = `
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 170 170" class="Progress Progress--medium">
  <g transform="rotate(90 85 85)">
    <circle stroke-width="22" fill="none" cx="85" cy="85" r="70" class="Progress-track"></circle>
  </g>
  <g transform="rotate(90 85 85)">
    <circle stroke-width="22" stroke-dasharray="0, 439.82259999999997" stroke-linecap="butt" fill="none" cx="85" cy="85" r="70" class="Progress-bar Progress-bar--green"></circle>
  </g> <text x="86" y="106" text-anchor="middle" class="Progress-label">

  </text>
</svg>
`;

exports[`VHProgressDonut.vue renders modified VHProgressDonut blue large 1`] = `
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 170 170" class="Progress Progress--large">
  <g transform="rotate(90 85 85)">
    <circle stroke-width="22" fill="none" cx="85" cy="85" r="70" class="Progress-track"></circle>
  </g>
  <g transform="rotate(90 85 85)">
    <circle stroke-width="22" stroke-dasharray="0, 439.82259999999997" stroke-linecap="butt" fill="none" cx="85" cy="85" r="70" class="Progress-bar Progress-bar--blue"></circle>
  </g> <text x="86" y="106" text-anchor="middle" class="Progress-label">

  </text>
</svg>
`;

exports[`VHProgressDonut.vue renders modified VHProgressDonut blue small 1`] = `
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 170 170" class="Progress Progress--medium">
  <g transform="rotate(90 85 85)">
    <circle stroke-width="22" fill="none" cx="85" cy="85" r="70" class="Progress-track"></circle>
  </g>
  <g transform="rotate(90 85 85)">
    <circle stroke-width="22" stroke-dasharray="4.398225999999999, 439.82259999999997" stroke-linecap="butt" fill="none" cx="85" cy="85" r="70" class="Progress-bar Progress-bar--blue"></circle>
  </g> <text x="86" y="106" text-anchor="middle" class="Progress-label">

  </text>
</svg>
`;

exports[`VHProgressDonut.vue renders modified VHProgressDonut green small 1`] = `
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 170 170" class="Progress Progress--small">
  <g transform="rotate(90 85 85)">
    <circle stroke-width="22" fill="none" cx="85" cy="85" r="70" class="Progress-track"></circle>
  </g>
  <g transform="rotate(90 85 85)">
    <circle stroke-width="22" stroke-dasharray="43.98226, 439.82259999999997" stroke-linecap="butt" fill="none" cx="85" cy="85" r="70" class="Progress-bar Progress-bar--green"></circle>
  </g> <text x="86" y="106" text-anchor="middle" class="Progress-label">

  </text>
</svg>
`;
